File: geekbang/algorithm/source-code/3/lc236 二叉树的最近公共祖先.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @param {TreeNode} p
 * @param {TreeNode} q
 * @return {TreeNode}
 */
var lowestCommonAncestor = function(root, p, q) {
    var father = {};
    var calcFather = function(root) {
        if (root == null) return;
        if (root.left != null) {
            father[root.left.val] = root;
            calcFather(root.left);
        }
        if (root.right != null) {
            father[root.right.val] = root;
            calcFather(root.right);
        }
    }

    calcFather(root);

    var redNodes = {}
    redNodes[root.val] = true;
    while (p.val != root.val) {
        redNodes[p.val] = true;
        p = father[p.val];
    }
    while (!redNodes[q.val]) {
        q = father[q.val];
    }
    return q;   
};

File: geekbang/algorithm/source-code/3/lc236 二叉树的最近公共祖先.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
        father = new HashMap<Integer, TreeNode>();
        calcFather(root);
        redNodes = new HashSet<Integer>();
        redNodes.add(root.val);
        while (p.val != root.val) {
            redNodes.add(p.val);
            p = father.get(p.val);
        }
        while (!redNodes.contains(q.val)) {
            q = father.get(q.val);
        }
        return q;
    }

    private void calcFather(TreeNode root) {
        if (root == null) return;
        if (root.left != null) {
            father.put(root.left.val, root);
            calcFather(root.left);
        }
        if (root.right != null) {
            father.put(root.right.val, root);
            calcFather(root.right);
        }
    }

    private Map<Integer, TreeNode> father;
    private Set<Integer> redNodes;
}